compiler/rustc_const_eval/src/check_consts/check.rs

```diff
@@ -773,39 +789,52 @@ impl<'tcx> Visitor<'tcx> for Checker<'_, 'tcx> {
                         let callee_safe_to_expose_on_stable =
                             is_safe_to_expose_on_stable_const_fn(tcx, callee);
 
                         // We only honor `span.allows_unstable` aka `#[allow_internal_unstable]` if
                         // the callee is safe to expose, to avoid bypassing recursive stability.
-                        if (self.span.allows_unstable(feature)
-                            || implied_feature.is_some_and(|f| self.span.allows_unstable(f)))
-                            && callee_safe_to_expose_on_stable
-                        {
+                        let is_allowed_unstable = |u: &rustc_attr::Unstability| {
+                            callee_safe_to_expose_on_stable
+                                && (self.span.allows_unstable(u.feature)
+                                    || u.implied_by.is_some_and(|f| self.span.allows_unstable(f)))
+                        };
+                        let mut needs_check =
+                            unstables.iter().filter(|u| !is_allowed_unstable(u)).peekable();
+                        if needs_check.peek().is_none() {
                             return;
                         }
 
                         // We can't use `check_op` to check whether the feature is enabled because
                         // the logic is a bit different than elsewhere: local functions don't need
                         // the feature gate, and there might be an "implied" gate that also suffices
                         // to allow this.
-                        let feature_enabled = callee.is_local()
-                            || tcx.features().enabled(feature)
-                            || implied_feature.is_some_and(|f| tcx.features().enabled(f));
+                        let is_feature_enabled = |u: &rustc_attr::Unstability| {
+                            callee.is_local()
+                                || tcx.features().enabled(u.feature)
+                                || u.implied_by.is_some_and(|f| tcx.features().enabled(f))
+                        };
                         // We do *not* honor this if we are in the "danger zone": we have to enforce
                         // recursive const-stability and the callee is not safe-to-expose. In that
                         // case we need `check_op` to do the check.
                         let danger_zone = !callee_safe_to_expose_on_stable
                             && self.enforce_recursive_const_stability();
-                        if danger_zone || !feature_enabled {
+                        let missing_features: SmallVec<[_; 1]> = if danger_zone {
+                            needs_check.map(|u| u.into()).collect()
+                        } else {
+                            needs_check
+                                .filter(|u| !is_feature_enabled(u))
+                                .map(|u| u.into())
+                                .collect()
+                        };
```

Comment:
This looks like it changes the logic -- now the closure defined up here also gets used below.

I'd prefer if you didn't change the logic but just did the "obvious" variant to generalize to multiple gates. Otherwise this will have to go somewhere on my pretty big "needs more thorough review" pile.

Reply:
I think I've fixed this. It should be the obvious generalization of the previous logic now.

Reply:
Yes it looks quite clear now, thanks. :)